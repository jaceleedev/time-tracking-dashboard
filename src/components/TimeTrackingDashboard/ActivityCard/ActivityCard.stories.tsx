import type { Meta, StoryObj } from '@storybook/react';
import ActivityCard from './ActivityCard';

const meta = {
  component: ActivityCard,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    backgroundColor: {
      description: 'Background color class for the card',
      control: 'text',
    },
    icon: {
      description: 'Icon configuration for the activity',
    },
    activity: {
      description: 'Activity data including title and timeframes',
    },
    selectedTimeframe: {
      description: 'Currently selected timeframe',
      control: 'select',
      options: ['daily', 'weekly', 'monthly'],
    },
  },
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof ActivityCard>;

export default meta;
type Story = StoryObj<typeof meta>;

const baseActivity = {
  timeframes: {
    daily: { current: 5, previous: 7 },
    weekly: { current: 32, previous: 36 },
    monthly: { current: 103, previous: 128 },
  },
};

export const Work: Story = {
  args: {
    backgroundColor: 'bg-light-orange',
    icon: {
      src: '/images/icon-work.svg',
      width: 79,
      height: 68,
      position: {
        right: 'right-[17.42px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Work',
    },
    selectedTimeframe: 'weekly',
  },
};

export const Play: Story = {
  args: {
    backgroundColor: 'bg-soft-blue',
    icon: {
      src: '/images/icon-play.svg',
      width: 90,
      height: 76,
      position: {
        right: 'right-[11px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Play',
      timeframes: {
        ...baseActivity.timeframes,
        weekly: { current: 10, previous: 8 },
      },
    },
    selectedTimeframe: 'weekly',
  },
};

export const Study: Story = {
  args: {
    backgroundColor: 'bg-light-red',
    icon: {
      src: '/images/icon-study.svg',
      width: 79,
      height: 72,
      position: {
        right: 'right-[17.07px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Study',
      timeframes: {
        ...baseActivity.timeframes,
        weekly: { current: 4, previous: 7 },
      },
    },
    selectedTimeframe: 'weekly',
  },
};

export const Exercise: Story = {
  args: {
    backgroundColor: 'bg-lime-green',
    icon: {
      src: '/images/icon-exercise.svg',
      width: 81,
      height: 54,
      position: {
        right: 'right-[17.62px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Exercise',
      timeframes: {
        ...baseActivity.timeframes,
        weekly: { current: 4, previous: 5 },
      },
    },
    selectedTimeframe: 'weekly',
  },
};

export const Social: Story = {
  args: {
    backgroundColor: 'bg-violet',
    icon: {
      src: '/images/icon-social.svg',
      width: 75,
      height: 85,
      position: {
        right: 'right-[13.48px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Social',
      timeframes: {
        ...baseActivity.timeframes,
        weekly: { current: 5, previous: 10 },
      },
    },
    selectedTimeframe: 'weekly',
  },
};

export const SelfCare: Story = {
  args: {
    backgroundColor: 'bg-soft-yellow',
    icon: {
      src: '/images/icon-self-care.svg',
      width: 67,
      height: 56,
      position: {
        right: 'right-[13.48px]',
      },
    },
    activity: {
      ...baseActivity,
      title: 'Self Care',
      timeframes: {
        ...baseActivity.timeframes,
        weekly: { current: 2, previous: 2 },
      },
    },
    selectedTimeframe: 'weekly',
  },
};
