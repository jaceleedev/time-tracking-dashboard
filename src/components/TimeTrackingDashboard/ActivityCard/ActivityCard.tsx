import Image from 'next/image';
import { useState } from 'react';

interface IconConfig {
  src: string;
  width: number;
  height: number;
  position: {
    right: string;
  };
}

interface TimeRecord {
  current: number;
  previous: number;
}

interface ActivityRecord {
  title: string;
  timeframes: {
    daily: TimeRecord;
    weekly: TimeRecord;
    monthly: TimeRecord;
  };
}

type TimeframeOption = 'daily' | 'weekly' | 'monthly';

interface ActivityCardProps {
  backgroundColor: string;
  icon: IconConfig;
  activity: ActivityRecord;
  selectedTimeframe: TimeframeOption;
}

const timeFrameLabels = {
  daily: 'Yesterday',
  weekly: 'Last Week',
  monthly: 'Last Month',
};

function ActivityCard({
  backgroundColor,
  icon,
  activity,
  selectedTimeframe,
}: Readonly<ActivityCardProps>) {
  const [isHovering, setIsHovering] = useState(false);
  const [isEllipsisHovering, setIsEllipsisHovering] = useState(false);

  const { title, timeframes } = activity;
  const { current, previous } = timeframes[selectedTimeframe];
  const previousLabel = timeFrameLabels[selectedTimeframe];

  const handleHover = (target: 'card' | 'ellipsis', hovering: boolean) => {
    if (target === 'card') {
      setIsHovering(hovering);
    } else {
      setIsEllipsisHovering(hovering);
    }
  };

  const cardClassName = `grid grid-cols-1 grid-rows-[22px_auto] gap-[6px] h-[122px] px-6 py-7 rounded-[15px] bg-dark-blue hover:cursor-pointer transition-colors lg:gap-y-6 lg:h-[199px] lg:px-[30px] lg:pt-[29px] ${isHovering && !isEllipsisHovering ? 'bg-twilight-blue' : ''}`;

  return (
    <article
      className={`w-full h-40 rounded-[15px] ${backgroundColor} lg:h-[244px]`}
    >
      <div className="relative h-[38px] overflow-hidden lg:h-[45px]">
        <Image
          src={icon.src}
          alt=""
          width={icon.width}
          height={icon.height}
          priority
          className={`absolute ${icon.position.right}`}
        />
      </div>
      <div
        onMouseEnter={() => handleHover('card', true)}
        onMouseLeave={() => handleHover('card', false)}
        aria-label={`${title} activity details`}
        className={cardClassName}
      >
        <header className="flex justify-between items-center relative z-10">
          <h3 className="text-card-title text-white">{title}</h3>
          <button
            onMouseEnter={() => handleHover('ellipsis', true)}
            onMouseLeave={() => handleHover('ellipsis', false)}
            aria-label="More options"
          >
            <Image
              src={'/images/icon-ellipsis.svg'}
              alt=""
              width={21}
              height={5}
            />
          </button>
        </header>
        <div className="flex justify-between items-center lg:flex-col lg:items-start lg:gap-y-2">
          <p className="text-hours text-white">{current}hrs</p>
          <p className="text-previous text-pale-blue">
            {previousLabel} - {previous}hrs
          </p>
        </div>
      </div>
    </article>
  );
}

export default ActivityCard;
