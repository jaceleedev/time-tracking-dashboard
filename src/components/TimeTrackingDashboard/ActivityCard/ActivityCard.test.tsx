import { describe, afterEach, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import ActivityCard from './ActivityCard';

expect.extend(toHaveNoViolations);

const mockIcon = {
  src: '/images/icon-work.svg',
  width: 79,
  height: 78,
  position: {
    right: 'right-[17.42px]',
  },
};

describe('ActivityCard', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  // 1. Rendering test
  it('renders correctly', () => {
    const mockActivity = {
      title: 'Work',
      timeframes: {
        daily: { current: 5, previous: 7 },
        weekly: { current: 32, previous: 36 },
        monthly: { current: 103, previous: 128 },
      },
    };
    render(
      <ActivityCard
        backgroundColor="bg-light-orange"
        icon={mockIcon}
        activity={mockActivity}
        selectedTimeframe="weekly"
      />
    );
    const title = screen.getByText('Work');
    const hours = screen.getByText('32hrs');
    const previousHours = screen.getByText('Last Week - 36hrs');
    expect(title).toBeDefined();
    expect(hours).toBeDefined();
    expect(previousHours).toBeDefined();
  });

  // 2. Props test
  it('displays correct data based on selectedTimeframe', () => {
    const mockActivity = {
      title: 'Work',
      timeframes: {
        daily: { current: 6, previous: 8 },
        weekly: { current: 33, previous: 37 },
        monthly: { current: 104, previous: 129 },
      },
    };
    render(
      <ActivityCard
        backgroundColor="bg-light-orange"
        icon={mockIcon}
        activity={mockActivity}
        selectedTimeframe="daily"
      />
    );
    const hours = screen.getByText('6hrs');
    const previousHours = screen.getByText('Yesterday - 8hrs');
    expect(hours).toBeDefined();
    expect(previousHours).toBeDefined();
  });

  // 3. Styles test
  it('applies the correct background color', () => {
    const mockActivity = {
      title: 'Work',
      timeframes: {
        daily: { current: 7, previous: 9 },
        weekly: { current: 34, previous: 38 },
        monthly: { current: 105, previous: 130 },
      },
    };
    const { container } = render(
      <ActivityCard
        backgroundColor="bg-light-orange"
        icon={mockIcon}
        activity={mockActivity}
        selectedTimeframe="weekly"
      />
    );
    const article = container.firstElementChild!;
    const icon = container.querySelector('img')!;
    expect(article.className).toContain('bg-light-orange');
    expect(icon).toHaveProperty('src');
    expect(icon.className).toContain('absolute');
  });

  // 4. Accessibility test
  it('has no accessibility violations', async () => {
    const mockActivity = {
      title: 'Work',
      timeframes: {
        daily: { current: 8, previous: 10 },
        weekly: { current: 35, previous: 39 },
        monthly: { current: 106, previous: 131 },
      },
    };
    const { container } = render(
      <ActivityCard
        backgroundColor="bg-light-orange"
        icon={mockIcon}
        activity={mockActivity}
        selectedTimeframe="weekly"
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 5. Snapshot test
  it('matches snapshot', () => {
    const mockActivity = {
      title: 'Work',
      timeframes: {
        daily: { current: 9, previous: 11 },
        weekly: { current: 36, previous: 40 },
        monthly: { current: 107, previous: 132 },
      },
    };
    const { asFragment } = render(
      <ActivityCard
        backgroundColor="bg-light-orange"
        icon={mockIcon}
        activity={mockActivity}
        selectedTimeframe="weekly"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
