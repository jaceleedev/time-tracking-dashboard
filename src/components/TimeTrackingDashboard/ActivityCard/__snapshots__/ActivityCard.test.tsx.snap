// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityCard > matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="w-full h-40 rounded-[15px] bg-light-orange"
  >
    <div
      class="relative h-[38px] overflow-hidden"
    >
      <img
        alt=""
        class="absolute right-[17.42px]"
        data-nimg="1"
        decoding="async"
        height="78"
        loading="lazy"
        src="/images/icon-work.svg"
        style="color: transparent;"
        width="79"
      />
    </div>
    <div
      class="grid grid-cols-1 grid-rows-[22px_38px] gap-[6px] h-[122px] px-6 py-7 rounded-[15px] bg-dark-blue"
    >
      <div
        class="flex justify-between items-center"
      >
        <h3
          class="text-card-title text-white"
        >
          Work
        </h3>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="5"
          loading="lazy"
          src="/images/icon-ellipsis.svg"
          style="color: transparent;"
          width="21"
        />
      </div>
      <div
        class="flex justify-between items-center"
      >
        <p
          class="text-hours text-white"
        >
          36hrs
        </p>
        <p
          class="text-previous text-pale-blue"
        >
          Last Week - 40hrs
        </p>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
